Document working with association and checkboxes in a form (#37885)
rails/rails: guides/source/form_helpers.md

NOTE: Pairs passed to `options_for_select` should have the text first and the value second, however with `options_from_collection_for_select` should have the value method first and the text method second.

> 

Reviewer: Was adding this extra line intentional?

> 

Author: ```suggestion
```

will produce the same output and the value chosen by the user can be retrieved by `params[:date][:year]`.

> Creating Checkboxes for Has Many relations

Reviewer: ```suggestion
Creating Checkboxes for Relations
```

------------------------------------------

> Generating checkboxes options that act appropriately with Active Record models and strong parameters can be a confusing undertaking. Luckily there is a helper function to significantly reduce the work required.

Reviewer: All new guides should be wrapped to 80 characters.

Author: I text wrapped the text but not the code blocks, I imagine this is what is desired?

> Here is what the markup may look like:

Reviewer: ```suggestion
Here is what the markup may look like:

```

```html
<input type="checkbox" value="1" checked="checked" name="person[city_ids][]" id="person_cities_1" />
<label for="person_cities_1">Pittsburgh</label>
<input type="checkbox" value="2" checked="checked" name="person[city_ids][]" id="person_cities_2"/>
<label for="person_cities_2">Madison</label>
<input type="checkbox" value="3" checked="checked" name="person[city_ids][]" id="person_cities_3"/>
<label for="person_cities_3">Santa Rosa</label>
```

You have a list of cities whose names are shown to the user and associated checkboxes that hold the id of each cities database entry.

### The Collection Check Boxes tag

The `collection_check_boxes` exists to help create these checkboxes with the name for each input that maps to params that can be passed directly to model relationships.

In this example we show the currently associated records in a collection:

```erb
<%= form_with model: @person do |person_form| %>
  <%= person_form.collection_check_boxes :city_ids, person.object.cities, :id, :name %>
<% end %>
```

> This would return

Reviewer: ```suggestion
This returns
```

```html
<input type="checkbox" value="1" checked="checked" name="person[city_ids][]" id="person_cities_1" checked="checked" />
<label for="person_cities_1">Pittsburgh</label>
<input type="checkbox" value="2" checked="checked" name="person[city_ids][]" id="person_cities_2" checked="checked" />
<label for="person_cities_2">Madison</label>
<input type="checkbox" value="3" checked="checked" name="person[city_ids][]" id="person_cities_3" checked="checked" />
<label for="person_cities_3">Santa Rosa</label>
```

You can also pass a block to format the html as you'd like:

```erb
<%= form_with model: @person do |person_form| %>
  <%= person_form.collection_check_boxes :city_ids, City.all, :id, :name do |city| %>
    <div>
      <%= city.check_box %><%= city.label %>
    </div>
  <% end %>
<% end %>
```

> This would return

Reviewer: ```suggestion
This returns
```

```html
<div>
  <input type="checkbox" value="1" checked="checked" name="person[city_ids][]" id="person_cities_1" checked="checked" />
  <label for="person_cities_1">Pittsburgh</label>
</div>
<div>
  <input type="checkbox" value="2" checked="checked" name="person[city_ids][]" id="person_cities_2" checked="checked" />
  <label for="person_cities_2">Madison</label>
</div>
<div>
  <input type="checkbox" value="3" checked="checked" name="person[city_ids][]" id="person_cities_3" checked="checked" />
  <label for="person_cities_3">Santa Rosa</label>
</div>
```

> The method signature is

Reviewer: ```suggestion
The method signature is

```

```erb
collection_check_boxes(object, method, collection, value_method, text_method, options = {}, html_options = {}, &block)
```
* `object` - the form object, this is unnecessary if using the `form.collection_check_boxes` method call
* `method` -  the association method for the model relation
* `collection` - an array of objects to show in checkboxes
* `value_method` -  the method to call when populating the value attribute in the checkbox

> * `text` - method is the method to call when populating the text of the label

Reviewer 2: @gwincr11 IMO `method is` is not necessary here if we see the earlier line.

Author: Good catch is a good catch that is 😅

will produce the same output and the value chosen by the user can be retrieved by `params[:date][:year]`.

Creating Checkboxes for Has Many relations
------------------------------------------

> Generating checkboxes options that act appropriately with ActiveRecord models and strong parameters by hand can be a confusing undertaking. Luckily there is a helper function to significantly reduce the work required.

Reviewer: ```suggestion
Generating checkboxes options that act appropriately with Active Record models and strong parameters can be a confusing undertaking. Luckily there is a helper function to significantly reduce the work required.
```

Here is what the markup may look like:
```html
<input type="checkbox" value="1" checked="checked" name="person[city_ids][]" id="person_cities_1" />
<label for="person_cities_1">Pittsburgh</label>
<input type="checkbox" value="2" checked="checked" name="person[city_ids][]" id="person_cities_2"/>
<label for="person_cities_2">Madison</label>
<input type="checkbox" value="3" checked="checked" name="person[city_ids][]" id="person_cities_3"/>
<label for="person_cities_3">Santa Rosa</label>
```

> You have a list of cities who's names are shown to the user and associated checkboxes that hold the id of each cities db entry.

Reviewer: ```suggestion
You have a list of cities whose names are shown to the user and associated checkboxes that hold the id of each cities database entry.
```

### The Collection Check Boxes tag

The `collection_check_boxes` exists to help create these checkboxes with the name for each input that maps to params that can be passed directly to model relationships.

In this example we show the currently associated records in a collection:

```erb
<%= form_with model: @person do |person_form| %>
  <%= person_form.collection_check_boxes :city_ids, person.object.cities, :id, :name %>
<% end %>
```

This would return

```html
<input type="checkbox" value="1" checked="checked" name="person[city_ids][]" id="person_cities_1" checked="checked" />
<label for="person_cities_1">Pittsburgh</label>
<input type="checkbox" value="2" checked="checked" name="person[city_ids][]" id="person_cities_2" checked="checked" />
<label for="person_cities_2">Madison</label>
<input type="checkbox" value="3" checked="checked" name="person[city_ids][]" id="person_cities_3" checked="checked" />
<label for="person_cities_3">Santa Rosa</label>
```

> You can also pass a block to formate the html as you feel fit:

Reviewer: ```suggestion
You can also pass a block to format the html as you'd like:
```

    <div>
      <%= city.check_box %><%= city.label %>
    </div>
  <% end %>
<% end %>
```

This would return

```html
<div>
  <input type="checkbox" value="1" checked="checked" name="person[city_ids][]" id="person_cities_1" checked="checked" />
  <label for="person_cities_1">Pittsburgh</label>
</div>
<div>
  <input type="checkbox" value="2" checked="checked" name="person[city_ids][]" id="person_cities_2" checked="checked" />
  <label for="person_cities_2">Madison</label>
</div>
<div>
  <input type="checkbox" value="3" checked="checked" name="person[city_ids][]" id="person_cities_3" checked="checked" />
  <label for="person_cities_3">Santa Rosa</label>
</div>
```

The method signature is
```erb
collection_check_boxes(object, method, collection, value_method, text_method, options = {}, html_options = {}, &block)
```
object it the form object, this is unnecessary if using the `form.collection_check_boxes` method call.
method is the association method for the model relation.
collection is an array of objects to show in checkboxes.
value_method is the method to call when populating the value attribute in the checkbox.

> text method is the method to call when populating the text of the label.

Reviewer: I think this would read better as:

```
* `object` - the form object, this is unnecessary if using the `form.collection_check_boxes` method call
* `method` -  the association method for the model relation
* `collection` - an array of objects to show in checkboxes
* `value_method` -  the method to call when populating the value attribute in the checkbox
* `text` - method is the method to call when populating the text of the label
```
